.buildkite/pipeline-nightly-e2e-tests.yml

```diff
@@ -0,0 +1,348 @@
+agents:
+  memory: "4G"
+
+steps:
+
+  - group: "aks"
+    steps:
+
+      - label: "aks"
+        commands:
+          - .ci/setenvconfig e2e/main
+          - .ci/setenvconfig e2e/aks
+          - .buildkite/scripts/common/get-test-artifacts.sh
+          - make run-deployer
+          - make e2e-run
```

Comment:
Could yaml anchors help to avoid these repetitive sections?  Also wondering if you could set a variable to template all of this in a yaml anchor?

Reply:
There is clearly a lot of repetition in this pipeline. The main interesting new feature is that we have almost all e2e-tests k8s distributions and versions declared in a single file. Not all because some versions are maintained directly in [hack/deployer/config/plans.yml](https://github.com/elastic/cloud-on-k8s/blob/main/hack/deployer/config/plans.yml).
Not that in Jenkins, we also have a lot of repetition between all e2e-tests pipeline.

That being said, we can generate this pipeline definition. I thought about it. I hesitated to bring this little extra complexity right away. Here is what it could look like using a go script, a go template and the list of k8s distributions to test in a JSON file:

<details>
<summary>

```go
> cd .buildkite/scripts/e2e-tests && \
  go run generate.go -template pipeline.template.yml -data environments.json
```

</summary>

```json
{
  "environments": [
    {
      "group": "aks",
      "versions": [
        "e2e/aks"
      ]
    },
    {
      "group": "eks",
      "versions": [
        "e2e/eks"
      ]
    },
    {
      "group": "eks-arm",
      "versions": [
        "e2e/eks-arm"
      ]
    },
    {
      "group": "gke",
      "versions": [
        "e2e/gke-k8s-versions 1.21",
        "e2e/gke-k8s-versions 1.22",
        "e2e/gke-k8s-versions 1.23"
      ]
    },
    {
      "group": "kind",
      "versions": [
        "e2e/kind-k8s-versions kindest/node:v1.20.15@sha256:6f2d011dffe182bad80b85f6c00e8ca9d86b5b8922cdf433d53575c4c5212248 0.14.0 ipv4",
        "e2e/kind-k8s-versions kindest/node:v1.21.12@sha256:f316b33dd88f8196379f38feb80545ef3ed44d9197dca1bfd48bcb1583210207 0.14.0 ipv4",
        "e2e/kind-k8s-versions kindest/node:v1.22.9@sha256:8135260b959dfe320206eb36b3aeda9cffcb262f4b44cda6b33f7bb73f453105 0.14.0 ipv4",
        "e2e/kind-k8s-versions kindest/node:v1.22.9@sha256:8135260b959dfe320206eb36b3aeda9cffcb262f4b44cda6b33f7bb73f453105 0.14.0 ipv6",
        "e2e/kind-k8s-versions kindest/node:v1.23.6@sha256:b1fa224cc6c7ff32455e0b1fd9cbfd3d3bc87ecaa8fcb06961ed1afb3db0f9ae 0.14.0 ipv4",
        "e2e/kind-k8s-versions kindest/node:v1.23.6@sha256:b1fa224cc6c7ff32455e0b1fd9cbfd3d3bc87ecaa8fcb06961ed1afb3db0f9ae 0.14.0 ipv6",
        "e2e/kind-k8s-versions kindest/node:v1.24.1@sha256:11276ce4763d5600379c0104c4cb51bf7420c8c02937708e958ad68d08e0910c 0.14.0 ipv4",
        "e2e/kind-k8s-versions kindest/node:v1.24.1@sha256:11276ce4763d5600379c0104c4cb51bf7420c8c02937708e958ad68d08e0910c 0.14.0 ipv6"
      ]
    },
    {
      "group": "ocp",
      "versions": [
        "e2e/ocp 4.7.49",
        "e2e/ocp 4.8.50",
        "e2e/ocp 4.9.48",
        "e2e/ocp 4.10.34",
        "e2e/ocp 4.11.5"
      ]
    },
    {
      "group": "resilience",
      "versions": [
        "e2e/resilience"
      ]
    },
    {
      "group": "tanzu",
      "versions": [
        "e2e/tanzu"
      ]
    }
  ]
}
```

```yaml
agents:
  memory: "4G"

steps:
{{ range $env := .environments }}

  - group: "{{ $env.group }}"
    steps:
    {{ range $version := $env.versions }}
      {{- if eq $env.group "kind" }}
      - label: "{{ $version }}"
        agents:
          provider: "gcp"
            image: "family/elastic-buildkite-agent-ubuntu-2004-lts"
            machineType: "n1-standard-16"
        commands:
          - .ci/setenvconfig e2e/main
          - .ci/setenvconfig {{ $version }}
          - .buildkite/scripts/common/get-test-artifacts.sh
          - make -C .ci TARGET="run-deployer e2e-run" ci
        artifact_paths:
          - "eck-diagnostic-*.zip"
      {{- else }}
      - label: "{{ $version }}"
        commands:
          - .ci/setenvconfig e2e/main
          - .ci/setenvconfig {{ $version }}
          - .buildkite/scripts/common/get-test-artifacts.sh
          - make run-deployer
          - make e2e-run
        artifact_paths:
          - "eck-diagnostic-*.zip"          
      {{- end }}
    {{- end }}
    {{ if ne $env.group "kind" }}
      - wait: ~
        continue_on_failure: true
      
      {{ range $version := $env.versions }}
      - label: "cleanup"
        commands:
          - .ci/setenvconfig cleanup/{{ $env.group }}
          - make run-deployer
      {{- end }}
    {{- end }}
{{- end }}

```

```go
package main

import (
	"encoding/json"
	"flag"
	"html/template"
	"io/ioutil"
	"os"
	"log"
)

var (
	tplFilename string
	dataFilename string
)

func main() {
	flag.StringVar(&tplFilename, "template", "", "Template filename")
	flag.StringVar(&dataFilename, "data", "", "JSON data filename")
	flag.Parse()

	if tplFilename == "" {
		log.Fatalf("Template file not defined")
	}
	if dataFilename == "" {
		log.Fatalf("JSON data file not defined")
	}
	tplBytes, err := ioutil.ReadFile(tplFilename)
	if err != nil {
		log.Fatalf("Fail to read template file %s, err: %s", tplFilename, err)
	}
	dataBytes, err := ioutil.ReadFile(dataFilename)
	if err != nil {
		log.Fatalf("Fail to read JSON data file %s, err: %s" + dataFilename, err)
	}

	var data interface{}
	err = json.Unmarshal(dataBytes, &data)
	if err != nil {
		log.Fatalf("Fail to parse JSON data file, err:", err)
	}

	t, err := template.New("tpl").Parse(string(tplBytes))
	if err != nil {
		log.Fatalf("Fail to parse template, err: %s", err)
	}
	err = t.ExecuteTemplate(os.Stdout, "tpl", data)
	if err != nil {
		log.Fatalf("Fail to execute template, err: %s", err)
	}
}
```

</details>